### PR TITLE
feat: picture-in-picture

### LINKED ISSUE #152: [Feature Request] Picture in picture
It would be great if the app allowed picture in picture mode.
This could allow you to for example continue watching a show or movie in background while responding to text messages.

### LINKED ISSUE #389: Entering split-screen mode resets playback position
**Describe the bug**
If I use [split-screen mode](https://developer.android.com/guide/topics/large-screens/multi-window-support), when app switches from full-screen mode, playback position resets to where it was when playback started:
* If video was just opened, it resets to 00:00
* If I was previously watching video and I opened it from main window of findroid at XX:YY, when I switch to split-screen it resets to XX:YY

**To Reproduce**
1. Open findroid
2. Start video
3. Open "recent tasks" system UI
4. Choose "Split" action in context menu

**Expected behavior**
App switches to split-screen mode with position at the same spot it were before "recent tasks" is opened.  
If playback was active, it should continue. Otherwise, paused.

**Screenshots**
~~If applicable, add screenshots to help explain your problem.~~

**Device info (please complete the following information):**
 - Device: OnePlus 7 Pro
 - Android version: 12
 - App version: 0.11.1
 - Jellyfin version: 10.8.10

**Additional context**
~~Add any other context about the problem here.~~

### PR DESCRIPTION
**Pull request by cd16b**
**Add Picture in picture**

Works with ExoPlayer and PlayerMPV

- add a button to enter the picture in picture mode in the player UI
- add a toggle in player settings
- settings to default pip aspect ratio to 16:9
- `android:launchMode="singleTask"` prevent player to crash
- override `onNewIntent()` and handle the new video

**Known problems:**

- _Android 11 and lower_ have strange pip behavior (like VLC for Android) -> won't fix

Let me know of any issues

Closes #152
Closes #389 

| Player | Settings |
| --- | --- |
| image | Screenshot_20230618-185006_Findroid Debug |

### CODE REVIEW

**Comment by Yash-Garg** [comment]
@cd16b You can also check https://github.com/jarnedemeulemeester/findroid/pull/238, there is an existing open PR for PiP!

**Comment by cd16b** [comment]
Sorry, I'm new to Github, how can I create a PR to that PR? Is it possible to make a PR to that fork? Let me know.

**Comment by Yash-Garg** [comment]
I tested this PR on my phone, works pretty good!

File: app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt [hunk 1]
```diff
@@ -69,10 +75,21 @@ class PlayerActivity : BasePlayerActivity() {
...
+        binding.playerView.addOnLayoutChangeListener { v: View?, oldLeft: Int,
+                                                       oldTop: Int, oldRight: Int, oldBottom: Int, newLeft: Int, newTop:
+                                                       Int, newRight: Int, newBottom: Int ->
+            if (viewModel.player is MPVPlayer) {
+                binding.playerView.getGlobalVisibleRect(sourceRectHint)
+            } else {
+                binding.playerView.videoSurfaceView?.getGlobalVisibleRect(sourceRectHint)
+            }
+        }
+
```
Comment (jarnedemeulemeester) [inline app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt:87]: Does this actually do anything?

File: app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt [hunk 2]
```diff
@@ -188,4 +224,38 @@ class PlayerActivity : BasePlayerActivity() {
...
+            pictureInPicture()
+        }
+    }
+
+    private fun pipParams(): PictureInPictureParams.Builder {
+        val aspectRatio = if (viewModel.player is MPVPlayer) {
+            Rational(binding.playerView.width, binding.playerView.height)
+        } else {
+            binding.playerView.player?.videoSize?.let { Rational(it.width, it.height) }
+        }
```
Comment (jarnedemeulemeester) [inline app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt:239]: I was able to fix the aspect ratio when using mpv with some changes to `MPVPlayer` that I would like to push. 
Would you be able to give me access? Seems like this is a protected branch (not sure how those exactly work)

**Review by jarnedemeulemeester — CHANGES_REQUESTED** [review]
Works pretty good but I still have a few issues.
- When not using PiP (disabled) and going home. Upon returning the playback is stopped and the user is returned to the previous fragment. One would assume that the player would still be open but paused.
- When entering PiP this error is thrown (but it seems to work fine):
```
E/ActivityThread: Performing pause of activity that is not resumed: {dev.jdtech.jellyfin.debug/dev.jdtech.jellyfin.PlayerActivity}
    java.lang.RuntimeException: Performing pause of activity that is not resumed: {dev.jdtech.jellyfin.debug/dev.jdtech.jellyfin.PlayerActivity}
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:5117)
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:5082)
        at android.app.servertransaction.PauseActivityItem.execute(PauseActivityItem.java:48)
        at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:179)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2364)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:205)
        at android.os.Looper.loop(Looper.java:294)
        at android.app.ActivityThread.main(ActivityThread.java:8052)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:578)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:946)
```
- When in PiP mode and starting playback of something else the playback starts in PiP but the UI is still from the main player making it pretty much unusable.

File: app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt (hunk 2 above)
Comment (cd16b) [reply app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt:239]: I removed the protection, are you able to access now?

File: app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt (hunk 1 above)
Comment (cd16b) [reply app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt:87]: Yes, it's used for a smooth transition when using the pip button or in android 12+ (not implemented) when `.setAutoEnterEnable(true)`.

Comment (cd16b) [reply app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt:87]: MPV seems to work with `binding.playerView.videoSurfaceView?.getGlobalVisibleRect(sourceRectHint)` so I removed the ifelse

File: app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt (hunk 2 above)
Comment (jarnedemeulemeester) [reply app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt]: Yes I was able to push!

**Comment by cd16b** [comment]
- When not using PiP (disabled) and going home... > now work
- When entering PiP this error is thrown... > I don't know why, onResume() shoud be called after onStart()
- When in PiP mode and starting playback... > I can't reproduce that behavior, I've added `finishAndRemoveTask()` let me know if it works

File: app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt (hunk 1 above)
Comment (cd16b) [reply app/phone/src/main/java/dev/jdtech/jellyfin/PlayerActivity.kt]: Works only with exoplayer, with MPV I wasn't able to get the `videoSurfaceView`

**Comment by cd16b** [comment]
- removed GESTURE_EXCLUSION_AREA_BOTTOM because already implemented in another PR

**Comment by jarnedemeulemeester** [comment]
> When in PiP mode and starting playback... > I can't reproduce that behavior, I've added finishAndRemoveTask() let me know if it works

Changing `finish()` to `finishAndRemoveTask()` does not seem to have changed anything.
 I have added a video to show what happens

[Kooha-2023-02-15-10-37-17.webm](https://user-images.githubusercontent.com/32322857/218990984-c9179adb-7bd4-4706-aab2-b717a39d6157.webm)

On Android 12 it seems fine, when starting a new item it opens the player in full screen.
But on Android 13 the new playback starts in PiP with broken controls as shown in the video.

BTW I noticed you created the PiP icons yourself and submitted them to Lucide icons, awesome!

**Comment by cd16b** [comment]
I think it's a bug of the Emulator, I tested on two Samsung android 13 device and it works fine

**Comment by jarnedemeulemeester** [comment]
Hmm strange. The bug does also happen on my Pixel 6 running Android 14 DP1. That's also where I first noticed the bug.
I don't quite remember if it also happened when running Android 13 but I would think so.

Here is what I propose. I will fix the linting issues tomorrow so that the CI succeeds and there are APK files to test with. 
Once that's done I will ask some people on the Discord server to test for this specific bug.

**Review by jarnedemeulemeester — APPROVED** [review]

**Comment by jarnedemeulemeester** [comment]
@cd16b I did a refactor of the pip implementation.
I also removed the options to completely disable pip and the pip button. I feel like the pip button should always be available when the device has support for it.

The previous bug that makes the app unusable when starting a new video is still present. I also have not found a solution for this.

Do you think we can merge this considering the bug?

**Comment by Yash-Garg** [comment]
Just to add here, I faced a crash with a video file when tried to switch into PiP mode. Adding stacktrace below:

```java
FATAL EXCEPTION: main
Process: dev.jdtech.jellyfin.debug, PID: 5046
java.lang.IllegalArgumentException: enterPictureInPictureMode: Aspect ratio is too extreme (must be between 0.418410 and 2.390000).
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3015)
	at android.os.Parcel.createException(Parcel.java:2995)
	at android.os.Parcel.readException(Parcel.java:2978)
	at android.os.Parcel.readException(Parcel.java:2920)
	at android.app.IActivityClientController$Stub$Proxy.enterPictureInPictureMode(IActivityClientController.java:1790)
	at android.app.ActivityClient.enterPictureInPictureMode(ActivityClient.java:344)
	at android.app.Activity.enterPictureInPictureMode(Activity.java:2908)
	at dev.jdtech.jellyfin.PlayerActivity.pictureInPicture(PlayerActivity.kt:254)
	at dev.jdtech.jellyfin.PlayerActivity.onCreate$lambda$6(PlayerActivity.kt:193)
	at dev.jdtech.jellyfin.PlayerActivity.$r8$lambda$4HTtrp33fm87PUsUaQPhle7TFFg(Unknown Source:0)
	at dev.jdtech.jellyfin.PlayerActivity$$ExternalSyntheticLambda8.onClick(Unknown Source:2)
	at android.view.View.performClick(View.java:7506)
	at android.view.View.performClickInternal(View.java:7483)
	at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0)
	at android.view.View$PerformClick.run(View.java:29334)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7872)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.wm.ActivityClientController.ensureValidPictureInPictureActivityParams(ActivityClientController.java:867)
	at com.android.server.wm.ActivityClientController.enterPictureInPictureMode(ActivityClientController.java:771)
	at android.app.IActivityClientController$Stub.onTransact(IActivityClientController.java:916)
	at com.android.server.wm.ActivityClientController.onTransact(ActivityClientController.java:125)
	at android.os.Binder.execTransactInternal(Binder.java:1280)
```

We should catch this exception and show a toast instead of straight away crashing the video stream.

File: app/phone/src/main/java/dev/jdtech/jellyfin/BasePlayerActivity.kt [hunk 3]
```diff
@@ -24,22 +26,40 @@ abstract class BasePlayerActivity : AppCompatActivity() {
...
+            viewModel.player.playWhenReady = false
+        }
     }
 
     override fun onStop() {
         super.onStop()
 
+        viewModel.playWhenReady = false
+        viewModel.player.playWhenReady = false
         mediaSession.release()
```
Comment (Yash-Garg) [inline app/phone/src/main/java/dev/jdtech/jellyfin/BasePlayerActivity.kt:53]: Adding this to `onStop()` fixed the issue that was happening with pip.

Can you try it out with this change? @jarnedemeulemeester 

```suggestion
		wasPip = false
        mediaSession.release()
```

**Review by Yash-Garg — CHANGES_REQUESTED** [review]

**Comment by cd16b** [comment]
I think it's better to wait, I'm trying to understand what has changed with Android 13, meanwhile let's see which devices have the bug

**Comment by cd16b** [comment]
Now it should work. I've tested it on the Android 13 emulator.

**Comment by travisterrell** [comment]
Is there any remaining thing that's preventing this merge from being completed? It appears the requester has fixed the reported issue a month ago, so just checking on the status. 

I'm happy to do some work if any is still required to get this out approved (testing/debugging/whatever), so if there's a real reason this is being held up, I'd be happy to help get that moving along. Otherwise, can we just get it completed?

Thanks in advance!

**Comment by cd16b** [comment]
The PR is ready to be merged.
One problem remains when you are in pip and play a new video; the video player is moved to a new task, so if you close the app, the video continues, and when you reopen the app, the video is shown in full screen rather than the main menu. But I couldn't fix it for now.
If you want to test it, let me know if there are other issues.

**Comment by travisterrell** [comment]
Great, thanks! I'll test it out. To be honest, the behavior you describe doesn't sound as bad as even Plex's PiP, ha. It's so rare for an app to have a perfect implementation, it seems. (YouTube and Hulu work pretty reliably, though.) 


I'll let you know if I encounter any major issues. Maybe having another tester will encourage them to approve it soon! 


Thanks, 

Travis 



On Apr 5, 2023 3:34 PM, Cd16d ***@***.***> wrote:

The PR is ready to be merged.
One problem remains when you are in pip and play a new video; the video player is moved to a new task, so if you close the app, the video continues, and when you reopen the app, the video is shown in full screen rather than the main menu. But I couldn't fix it for now.
If you want to test it, let me know if there are other issues.

**Comment by cd16b** [comment]
**fix:**
- aspect ratio is too extreme for vertical videos
- update PlayerItems and initialize player again instead of recreate Activity

**add:**
- settings to default pip aspect ratio to 16:9

**problems:**
- Android 11 and lower move pip activity to a new task -> strange pip behavior (like VLC for Android)

**Comment by cd16b** [comment]
fix #389

**Comment by Nadeem-05** [comment]
any update on this pr?

**Comment by jarnedemeulemeester** [comment]
@cd16b I have merged the latest changes from main into this branch and fixed a few things because of the merge

**Review by jarnedemeulemeester — COMMENTED** [review]
Everything seems to work well.
I just have one question :)

File: core/src/main/res/xml/fragment_settings_player.xml [hunk 4]
```diff
@@ -99,12 +99,23 @@
...
 
+    <PreferenceCategory app:title="@string/picture_in_picture">
+        <SwitchPreferenceCompat
+            app:key="pref_player_picture_in_picture_gesture"
+            app:title="@string/picture_in_picture_gesture"
+            app:summary="@string/picture_in_picture_gesture_summary" />
+        <SwitchPreferenceCompat
+            app:key="pref_player_picture_in_picture_aspect_ratio"
+            app:summary="@string/picture_in_picture_aspect_ratio_summary"
+            app:title="@string/picture_in_picture_aspect_ratio" />
```
[the remark this answers is not in the document]
Comment (cd16b) [reply core/src/main/res/xml/fragment_settings_player.xml:118]: I use it on an old Andorid phone, where the pip window cannot be resized, but we can remove it.

**Comment by travisterrell** [comment]
Thanks for the hard work and perseverance! Excited for this to be merged!

**Pull request merged**